sorting.go

```diff
@@ -0,0 +1,205 @@
+			SortingColumns: config.SortingColumns,
+		}),
+		buffer: buffer,
+		writer: NewGenericWriter[T](buffer, &WriterConfig{
+			CreatedBy:            config.CreatedBy,
+			ColumnPageBuffers:    config.ColumnPageBuffers,
+			ColumnIndexSizeLimit: config.ColumnIndexSizeLimit,
+			PageBufferSize:       config.PageBufferSize,
+			WriteBufferSize:      config.WriteBufferSize,
+			DataPageVersion:      config.DataPageVersion,
+			Schema:               config.Schema,
+			SortingColumns:       config.SortingColumns,
+			Compression:          config.Compression,
+		}),
+		output:  NewGenericWriter[T](output, config),
+		maxRows: sortRowCount,
+		sorting: config.SortingColumns,
+	}
+}
+
+func (w *SortingWriter[T]) Close() error {
+	if err := w.Flush(); err != nil {
+		return err
+	}
+	return w.output.Close()
+}
+
+func (w *SortingWriter[T]) Flush() error {
+	if err := w.sortAndWriteBufferedRows(); err != nil {
+		return err
+	}
+
+	if w.numRows == 0 {
+		return nil
+	}
+
+	defer func() {
+		w.buffer.Reset()
+		w.writer.Reset(w.buffer)
+		w.numRows = 0
+	}()
+
+	if err := w.writer.Close(); err != nil {
+		return err
+	}
+
+	f, err := OpenFile(bytes.NewReader(w.buffer.Bytes()), int64(w.buffer.Len()),
+		&FileConfig{
+			SkipPageIndex:    true,
+			SkipBloomFilters: true,
+			ReadBufferSize:   defaultReadBufferSize,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	m, err := MergeRowGroups(f.RowGroups(),
+		&RowGroupConfig{
+			Schema:         w.Schema(),
+			SortingColumns: w.sorting,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if _, err := w.output.WriteRowGroup(m); err != nil {
+		return err
+	}
+
+	return w.output.Flush()
+}
+
+func (w *SortingWriter[T]) Reset(output io.Writer) {
+	w.rows.Reset()
+	w.buffer.Reset()
+	w.writer.Reset(w.buffer)
+	w.output.Reset(output)
+	w.numRows = 0
+}
+
+func (w *SortingWriter[T]) Write(rows []T) (int, error) {
+	return w.writeRows(len(rows), func(i, j int) (int, error) { return w.rows.Write(rows[i:j]) })
+}
+
+func (w *SortingWriter[T]) WriteRows(rows []Row) (int, error) {
+	return w.writeRows(len(rows), func(i, j int) (int, error) { return w.rows.WriteRows(rows[i:j]) })
+}
+
+func (w *SortingWriter[T]) writeRows(numRows int, writeRows func(i, j int) (int, error)) (int, error) {
+	wn := 0
+
+	for wn < numRows {
+		if w.rows.NumRows() >= w.maxRows {
+			if err := w.sortAndWriteBufferedRows(); err != nil {
+				return wn, err
+			}
+		}
+
+		n := int(w.maxRows - w.rows.NumRows())
+		n += wn
+		if n > numRows {
+			n = numRows
+		}
+
+		n, err := writeRows(wn, n)
+		wn += n
+
+		if err != nil {
+			return wn, err
+		}
+	}
+
+	return wn, nil
+}
+
+func (w *SortingWriter[T]) SetKeyValueMetadata(key, value string) {
+	w.output.SetKeyValueMetadata(key, value)
+}
+
+func (w *SortingWriter[T]) Schema() *Schema {
+	return w.output.Schema()
+}
+
+func (w *SortingWriter[T]) sortAndWriteBufferedRows() error {
+	if w.rows.Len() == 0 {
+		return nil
+	}
+
+	defer w.rows.Reset()
+	sort.Sort(w.rows)
```

Comment:
Thoughts on writing to a heap instead of sorting here?

Reply:
Do you mean maintaining a heap on the buffered rows as they are written?

Reply:
Yup, so the cost of sorting is amortized across all of the writes instead of done once when it's time to flush.

Reply:
The heap is not sorted tho, it seems that it means we need to load all two groups back to memory when merging since we may have to manipulate the sub heaps as they are consumed by the merge, or am I missing something?

Reply:
My suggestion is likely not impactful unless `maxRows` is quite large.

As far as I can tell you are buffering rows into `w.rows` in the order they are written. Then, on this line, are performing a sort on the rows so that you can flush them in order to an in memory row group. Later these are merged to write to the output file.

If, instead of a slice, `RowBufffer[T]` was backed by a heap data structure they could be retrieved in order and the cost of sorting would be amortized across all of the row writes.

However, digging deeper, the `RowBuffer` struct implements the `RowGroup` interface which includes `SeekToRow` which would be difficult to implement if all of this was backed with a heap. So, you are welcome to ignore! This was idle thought while waking up with a coffee and on deeper inspection doesn't pan out.